fix: invalid hook call and limit events data fetching

bandhan-majumder opened the pull request:
**What kind of change does this PR introduce?**

bugfix

**Issue Number:**

Fixes #3179

**Snapshots/Videos:**

https://www.loom.com/share/a495245266f341e6a2e379daf4286265

<!--Add snapshots or videos wherever possible.-->

**If relevant, did you update the documentation?**

No

**Summary**

1. Fixes setup bug
2. Limits event data fetching
3. Fix invalid hook call sequence

**Does this PR introduce a breaking change?**

Not sure

## Checklist

### CodeRabbit AI Review
- [ ] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [ ] I have implemented or provided justification for each non-critical suggestion
- [ ] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [ ] I have written tests for all new changes/features
- [ ] I have verified that test coverage meets or exceeds 95%
- [ ] I have run the test suite locally and all tests pass


**Other information**

N/A

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

<!--Yes or No-->

bandhan-majumder commented:
PTAL @palisadoes . Failing as it's based on `develop` branch.

meetulr commented:
@bandhan-majumder This only fetches the first 400 events from the database; it’s not a fix for this issue. A more efficient approach for now would be to trim the sample data as a quick fix.

This issue will only be fully resolved when we modify the query itself. There’s a new discussion addressing this: [Recurring Events](https://github.com/PalisadoesFoundation/talawa/discussions/2763). PTAL.

palisadoes commented:
@bandhan-majumder Please work with @meetulr  on getting this PR approved

bandhan-majumder commented:
Yes @meetulr . As I also said earlier, it's a temporary fix. For a permanent fix, we may need to fix how it is retrieved. And handling the recurrence in the frontend will make the most sense. What do you think? Currently it's trying to get ~8080 arrays of events. I hard coded it to 400 to make it faster.

bandhan-majumder commented:
@meetulr one another thing we can think of doing is to fetch data based on the particular month or date or year. But that would make the logic unnecessarily complex for recurring events. Also there will be multiple calls based on how user navigates to see other events.

palisadoes commented:
@coderabbitai full review

bandhan-majumder commented on src/screens/UserPortal/Events/Events.tsx:
```diff
@@ -68,39 +69,64 @@
-  const [viewType, setViewType] = React.useState<ViewType>(ViewType.MONTH);
-  const [createEventModal, setCreateEventmodalisOpen] = React.useState(false);
-  const [createChatCheck, setCreateChatCheck] = React.useState(false);
+  const [events, setEvents] = useState([]);
+  const [eventTitle, setEventTitle] = useState('');
+  const [eventDescription, setEventDescription] = useState('');
+  const [eventLocation, setEventLocation] = useState('');
+  const [startDate, setStartDate] = useState<Date | null>(new Date());
+  const [endDate, setEndDate] = useState<Date | null>(new Date());
+  const [isPublic, setIsPublic] = useState(true);
+  const [isRegisterable, setIsRegisterable] = useState(true);
+  const [isRecurring, setIsRecurring] = useState(false);
+  const [isAllDay, setIsAllDay] = useState(true);
+  const [startTime, setStartTime] = useState('08:00:00');
+  const [endTime, setEndTime] = useState('10:00:00');
+  const [viewType, setViewType] = useState<ViewType>(ViewType.MONTH);
+  const [createEventModal, setCreateEventmodalisOpen] = useState(false);
+  const [createChatCheck, setCreateChatCheck] = useState(false);
```
It's already there

bandhan-majumder commented:
@palisadoes any feedback on this? Fixing it from the core may take a bit longer time. I had done changes to get only 400 event datas so that it becomes visible to the user quickly and they get a taste of what does it actually do. And it's the temporary fix as asked.

meetulr commented:
> Yes @meetulr . As I also said earlier, it's a temporary fix. For a permanent fix, we may need to fix how it is retrieved. And handling the recurrence in the frontend will make the most sense. What do you think? Currently it's trying to get ~8080 arrays of events. I hard coded it to 400 to make it faster.

It will not work when the first 400 events all ended prior to the current calendar date. Then the calendar will be empty as there were no entries fetched for that date.

> one another thing we can think of doing is to fetch data based on the particular month or date or year. But that would make the logic unnecessarily complex for recurring events. Also there will be multiple calls based on how user navigates to see other events.

Something has to be done to make it efficient. It will not be straightforward, will depend on the implementation of recurring events.

I think it should not be just a frontend issue. Either the sample data trimmer, or the query modification in both api and admin.

bandhan-majumder commented:
@meetulr 
Yes. API (maybe DB too) should be involved in fixing this issue efficiently. And you are right that the calendar may appear empty if 400 event ends prior. Will love to fix this issue permanently. But as @palisadoes said, the develop branch which is used in [demo.talawa.io](https://demo.talawa.io/) site, there user should get a taste of what the service looks like. I added the temp fix based on that only. User will get to see few events. We may reach to the permanent solution with the discussion going on about this after some time.

meetulr commented:
> @meetulr Yes. API (maybe DB too) should be involved in fixing this issue efficiently. And you are right that the calendar may appear empty if 400 event ends prior. Will love to fix this issue permanently. But as @palisadoes said, the develop branch which is used in [demo.talawa.io](https://demo.talawa.io/) site, there user should get a taste of what the service looks like. I added the temp fix based on that only. User will get to see few events. We may reach to the permanent solution with the discussion going on about this after some time.

If it's urgent, then ok.
But keep in mind there are too many sample events, 400 will run out soon, then the calendar will be empty, and even the newly created events won't be visible, and the app won't look functional.

That's why I said that a better temporary fix will be sample data trimming, until the query is modified. So we'll have to focus on that...
And that too will be temporary until we have a new implementation that'll require modifying the query to fit that approach.

bandhan-majumder commented:
@meetulr yes. But there are several ways to trim the data. For example;
1. getting the data of current month only
2. getting the data of 3 months (past, present, future)
and so on...

One another thing we can add is skipping the data dynamically where the 400 data will be taken dynamically after skipping X number of events. Do you think that will be better? With that approach maybe we will not run out of events any time soon.

meetulr commented:
> @meetulr yes. But there are several ways to trim the data. For example;
> 
> 1. getting the data of current month only (the other months data may get empty)
> 2. getting the data of 3 months (past, present, future)
>    and so on...
> 

By trimming the sample data, I simply meant reducing the number of documents. Other solutions will require modifying the query.

> One another thing we can add is skipping the data dynamically where the 400 data will be taken dynamically after skipping X number of events. Do you think that will be better? With that approach maybe we will not run out of events any time soon.

Yes, for now we can do something like fetching some events where the `startDate` is greater than the current month's (or the last month's) start date.

Solutions like these would all be temp fixes, but it will atleast keep the calendar filled and the app functional.

bandhan-majumder commented:
@meetulr @palisadoes 

There is a very easy, faster, simple and reasonable solution of this that I overlooked. There is already api support for this. We can fetch `upcomingOnly`  and limited `400` events. I have updated the video with the current implementation video to show how it looks after this change. This approach will load the events faster and will never run out of events.

meetulr commented:
@bandhan-majumder should be ok for now.

One question though, why is checkEnv modified?

bandhan-majumder commented:
@meetulr script was failing if there was no
 .env file beforehand. So I added a check that if it's not present, it will create a file `.env` so that it does not fail. And added tests to improve coverage.

meetulr commented:
> @meetulr script was failing if there was no .env file beforehand. So I added a check that if it's not present, it will create a file `.env` so that it does not fail. And added tests to improve coverage.

Could you elaborate on that? Is it necessary in this pr? If not it should be a separate issue.

bandhan-majumder commented:
here is what is solves: https://www.loom.com/share/e65087c80af44290832458ce871e8ff8
Not necessary. But I have told this in the [issue](https://github.com/PalisadoesFoundation/talawa-admin/issues/3179#issuecomment-2687053405). It's a simple bug and may not required another pr for another if-else. Please let me know.

meetulr commented:
Alright, it'll do for now. But separate issues should be handled separately to avoid confusion.

bandhan-majumder commented:
Sure, thanks. @palisadoes PTAL
